List Subcommand (Implementation) (#3523)
model-checking/kani: rfc/src/rfcs/0013-list.md

## User Experience

Users run a `list` subcommand, which prints metadata about the harnesses and contracts in each crate under verification. The subcommand takes two options:
- `--message-format=[pretty|json]`: change the output format. The default is `pretty`, which prints to the terminal. The `json` option creates and writes to a JSON file instead.
- `--std`: Include if we are running on the standard library. This option is only available for `kani list` (not `cargo kani list`), which mirrors the verification workflow for the standard library.

This subcommand does not fail. In the case that it does not find any harnesses or contracts, it prints a message informing the user of that fact.

### Pretty Format

The default format, `pretty`, prints a "Contracts" table and a "Standard Harnesses" list.

> Each row of the "Contracts" table consists of a function, the number of contracts it has, and its contract harnesses.

Author: Yes, my bad -- forgot to update this when we removed the contracts count.

> Each row of the "Contracts" table consists of a function, the number of contracts it has, and its contract harnesses.

Author: ```suggestion
Each row of the "Contracts" table consists of a function under contract and its contract harnesses.
```

> A function is listed if it has contracts or it is the target of contract harness(es).

Author: ```suggestion
```

|-------|-------------------------|----------------|--------------------------------------------------|
|       | example::impl::foo      | 2              | example::verify::check_foo_u32                   |
|       |                         |                | example::verify::check_foo_u64                   |
|-------|-------------------------|----------------|--------------------------------------------------|
|       | example::impl::func     | 1              | example::verify::check_func                      |
|-------|-------------------------|----------------|--------------------------------------------------|
|       | example::prep::parse    | 1              | NONE                                             |
|-------|-------------------------|----------------|--------------------------------------------------|
| Total | 5                       | 8              | 5                                                |
|-------|-------------------------|----------------|--------------------------------------------------|

Standard Harnesses (#[kani::proof]):
1. example::verify::check_modify

> 2. example::verify::check_new

Reviewer: Why a numbered list?

Author: Because it's a concise way to see how many standard harnesses you have--we can report the list and the total simultaneously.

Reviewer: I personally like the previous format better with the explicit total. But I'll leave it up to you. Thanks

## Software Design

### Metdata Changes
We introduce a new `ContractedFunction` struct to `kani_metadata`:
```rust

> pub struct ContractedFunction {

Reviewer: TBH, I don't think you need to add this level of detail. I think saying high-level that the metadata will now include a list of functions that contains contract and information about their harnesses.

    /// The fully qualified name the user gave to the function (i.e. includes the module path).

>     pub function: String,

Reviewer: This is per function definition not instantiation, right?

Author: Yes.

See [this blog post](https://model-checking.github.io/kani-verifier-blog/2022/10/27/using-kani-with-the-bolero-property-testing-framework.html) for more information.

> There's no easy way for us to know whether a harness comes from Bolero, since Bolero takes care of rewriting the test to use Kani syntax and invoking the Kani engine. By the time the harness gets to Kani, there's no way for us to tell it apart from a regular harness. Fixing this would require some changes to our Bolero integration.

Reviewer 2: I'm not sure I follow the comment here. Bolero harnesses use the `kani::proof` attribute just like any other proof. Is the aim to distinguish bolero harnesses from kani harnesses?

> There's no easy way for us to know whether a harness comes from Bolero, since Bolero takes care of rewriting the test to use Kani syntax and invoking the Kani engine. By the time the harness gets to Kani, there's no way for us to tell it apart from a regular harness. Fixing this would require some changes to our Bolero integration.

Reviewer 2: Ah I just expanded the paragraph above and that's exactly what the doc says:

> It would be nice to differentiate between regular Kani harnesses and Bolero harnesses. Bolero harnesses invoke Kani using conditional compilation

I'm open to any suggestions on how to improve this.

> There's no easy way for us to know whether a harness comes from Bolero, since Bolero takes care of rewriting the test to use Kani syntax and invoking the Kani engine. By the time the harness gets to Kani, there's no way for us to tell it apart from a regular harness. Fixing this would require some changes to our Bolero integration.

Reviewer: In theory, I think we can look at the list of functions reachable from the harness and determine if it uses bolero. I'm not sure the extra overhead is worth it though.

> There's no easy way for us to know whether a harness comes from Bolero, since Bolero takes care of rewriting the test to use Kani syntax and invoking the Kani engine. By the time the harness gets to Kani, there's no way for us to tell it apart from a regular harness. Fixing this would require some changes to our Bolero integration.

Reviewer 2: Yeah I'm not sure it's really worth distinguishing a bolero harness vs a vanilla Kani one... It's not super clear to me what one would actually do with that information.
